src/coredump/coredump.c

```diff
@@ -1229,6 +1245,50 @@ static int gather_pid_metadata(struct iovec_wrapper *iovw, Context *context) {
         return save_context(context, iovw);
 }
 
+static int process_kernel_in_namespace(const Context *context, int argc, char **argv) {
+        _cleanup_close_ int pidnsfd = -1, mntnsfd = -1, usernsfd = -1, rootfd = -1;
+        _cleanup_strv_free_ char **arguments = NULL;
+        pid_t child;
+        int r;
+
+        /* Invoke systemd-coredump in the container. We pass the same list of arguments
+         * the kernel passed us but we make sure to translate any sensitive metadata */
+
+        log_debug("Invoking %s in the container...", argv[0]);
+
+        /* Copy the argument list now because namespace_fork() is going to replace child's
+         * argv[] completely */
+        arguments = strv_copy(argv);
+        if (!arguments)
+                return log_oom();
+
+        r = namespace_open(context->pid, &pidnsfd, &mntnsfd, NULL, &usernsfd, &rootfd);
+        if (r < 0)
+                return log_error_errno(r, "Can't retrieve namespaces of crashing process: %m");
+
+        r = namespace_fork("(sd-coredumpns)", "(sd-coredump)", NULL, 0,
+                           FORK_RESET_SIGNALS|FORK_DEATHSIG,
+                           pidnsfd, mntnsfd, -1, usernsfd, rootfd, &child);
+        if (r < 0)
+                return r;
+        if (r == 0) {
+                /* Make sure to not leak PID as seen in the host pid namespace */
+                arguments[META_ARGV_PID] = arguments[META_ARGV_NS_PID];
+
+                (void) execv(arguments[0], arguments);
+
+                log_error_errno(errno, "Failed to execute '%s' in container: %m", arguments[0]);
+                _exit(EXIT_FAILURE);
```

Comment:
of course there should be a fallback: if systemd-coredump doesn't exist in the container we should process the coredump on the host, as before.

Also, systemd-coredump hands off the handling of the coredump immediately to a socket so that the main work runs as system service. This means that if we are in a container that includes the systemd binaries but is not booted with systemd might end up in a confused state where we invoke systemd-coredump in the container, but it cannot work since it can't talk to its socket.

We should add some safety for that. i think a simple access() check beforehand to see if the socket in /run exists should suffice: if it doesn't we should assume the container is not actually started and should then process the coredump on the host.

Also, containers and host might run different distros. iirc debian has different paths for /usr/lib/systemd-coredump (i.e. might use /lib rather than /usr/lib?) I think we should hence try execv() on the various possibly paths before giving up.

Reply:
Agreed with all points.